C error when using or keyword on a map property of a generic parameter

**What did you do?**

I ran this code:

```v
import vweb
import req

struct App {
  vweb.Context
}

fn (mut app App) index() vweb.Result {
  name := req.get_data(app, "name") or { "John" }
  // key := "name"
  // name := app.query[key] or { "John" }

  return app.html('<h1>Hello, $name</h1>')
}

fn main() {
  vweb.run(&App{}, 8000)
}
```

Which uses this code for the "req" local package

```v
module req

pub fn get_data<T>(app T, key string) ?string {
  value := app.query[key]

  if value == "" {
    return none
  }

  return app.query[key] or { none }
}
```

**What did you expect to see?**

The web server listens requests as expected

 
**What did you see instead?**

This erorr when running the web server

```bash
$ v run src/index.v 
==================
/tmp/v/index.290258594001423207.tmp.c: In function 'req__get_data_T_main__App':
/tmp/v/index.290258594001423207.tmp.c:11476:27: error: incompatible types when assigning to type 'string' {aka 'struct string'} from type 'IError' {aka 'struct <anonymous>'}
11476 |   *(string*) _t692.data = _const_none__;
      |                           ^~~~~~~~~~~~~
...
```

**Notes**

Thanks to @spytheman , we found that this code will compile correctly:

```v
import vweb
import req

struct App {
  vweb.Context
}

fn (mut app App) index() vweb.Result {
  // name := req.get_data(app, "name") or { "John" }
  key := "name"
  name := app.query[key] or { "John" }

  return app.html('<h1>Hello, $name</h1>')
}

fn main() {
  vweb.run(&App{}, 8000)
}
```

| OS                    | linux, Linux version 5.4.72-microsoft-standard-WSL2 (oe-user@oe-host) (gcc version 8.2.0 (GCC)) #1 SMP Wed Oct 28 23:40:43 UTC 2020 (WSL 2) |
|-----------------------|---------------------------------------------------------------------------------------------------------------------------------------------|
| Processor             | 12 cpus, 64bit, little endian, AMD Ryzen 5 3600 6-Core Processor                                                                            |
| CC Version            | cc (Alpine 9.3.0) 9.3.0                                                                                                                     |
| getwd                 | /home/alpine                                                                                                                                |
| vmodules              | /root/.vmodule                                                                                                                              |
| vroot                 | /opt/vlang                                                                                                                                  |
| vexe                  | /opt/vlang/v                                                                                                                                |
| vexe mtime            | 2021-05-22 14:07:39                                                                                                                         |
| is vroot writable     | true                                                                                                                                        |
| is vmodules writable  | true                                                                                                                                        |
| V full version        | V 0.2.2 788520e.e512caf                                                                                                                     |
| env VFLAGS            | "-cc gcc"                                                                                                                                   |
| Git version           | git version 2.24.4                                                                                                                          |
| Git vroot status      | weekly.2021.20-41-ge512caf8 (22 commit(s) behind V master)                                                                                  |
| .git/config present   | true                                                                                                                                        |
| thirdparty/tcc status | thirdparty-linux-amd64 7543de81                                                                                                             |

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -6521,6 +6521,7 @@ pub fn (mut c Checker) index_expr(mut node ast.IndexExpr) ast.Type {
 		}
 	}
 	c.stmts(node.or_expr.stmts)
+	c.check_expr_opt_call(node, typ)
 	return typ
 }
 

--- a/vlib/v/checker/tests/or_expr_types_mismatch.out
+++ b/vlib/v/checker/tests/or_expr_types_mismatch.out
@@ -1,0 +1,14 @@
+vlib/v/checker/tests/or_expr_types_mismatch.vv:3:19: error: wrong return type `none` in the `or {}` block, expected `string`
+    1 | fn get_map() ?string {
+    2 |     m := map{1: 'a', 2: 'b'}
+    3 |     return m[1] or { none }
+      |                      ~~~~
+    4 | }
+    5 |
+vlib/v/checker/tests/or_expr_types_mismatch.vv:8:19: error: wrong return type `none` in the `or {}` block, expected `int`
+    6 | fn get_array() ?int {
+    7 |     a := [1, 2, 3]
+    8 |     return a[4] or { none }
+      |                      ~~~~
+    9 | }
+   10 |

--- a/vlib/v/checker/tests/or_expr_types_mismatch.vv
+++ b/vlib/v/checker/tests/or_expr_types_mismatch.vv
@@ -1,0 +1,17 @@
+fn get_map() ?string {
+	m := map{1: 'a', 2: 'b'}
+	return m[1] or { none }
+}
+
+fn get_array() ?int {
+	a := [1, 2, 3]
+	return a[4] or { none }
+}
+
+fn main() {
+	map_result := get_map() or { return }
+	println(map_result)
+
+	array_result := get_array() or { return }
+	println(array_result)
+}

